feat(RingTheory/MvPolynomial/Groebner): add Gröbner basis

Hagb opened the pull request:
This PR adds some definitions and theorems of Gröbner basis theory.

Definitions:

- `MonomialOrder.IsRemainder`
- `MonomialOrder.IsGroebnerBasis`

Main theorems:


* `MonomialOrder.remainder_eq_zero_iff_mem_ideal_of_isGroebnerBasis`: Given a remainder of a polynomial on division by a Gröbner basis of an ideal, the remainder is 0 if and only if the polynomial is in the ideal.
* `MonomialOrder.isGroebnerBasis_iff_subset_ideal_and_isRemainder_zero`: A finite set of polynomials is a Gröbner basis of an ideal if and only if it is a subset of this ideal and 0 is a remainder of each member of this ideal on division by this finite set.
* `MonomialOrder.existsUnique_isRemainder_of_isGroebnerBasis`: Remainder of any polynomial on division by a Gröbner basis exists and is unique.
* `MonomialOrder.ideal_eq_span_of_isGroebnerBasis`: Gröbner basis of any ideal spans the ideal.
* `MonomialOrder.isGroebnerBasis_iff_isRemainder_sPolynomial_zero` (Buchberger Criterion): a basis of an ideal is a Gröbner basis of it if and only if 0 is a remainder of echo sPolynomial between two polynomials on the basis.

The previous `Mathlib/RingTheory/MvPolynomial/Groebner.lean` is moved to `Mathlib/RingTheory/MvPolynomial/Groebner/Division.lean`, and used by `Mathlib/RingTheory/MvPolynomial/Groebner/Remainder.lean` for the proof of remainder's existence.

The PR is upstreamized from https://github.com/WuProver/groebner_proj.

Co-authored-by: Hao Shen [3118181069@qq.com](mailto:3118181069@qq.com)

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

CC @tsuki8

TODO / WIP:

- [x] clean up some proofs
- [x] split lemmas
- [x] squeeze Non-terminal `simp`s
- [x] add comments to explain the proof of Buchberger Criterion.
- [x] refactor something with `WithBot (σ →₀ ℕ)` variant of `MonomialOrder.degree` (`MonomialOrder.withBotDegree`, #34759).
- [x] refactor to use namespaces `MonomialOrder.{IsGroebnerBasis,IsRemainder}` instead of `*_of_{isGroebnerBasis,isRemainder}` in theorem names.
- [ ] reduced Groebner Basis (done, not yet submitted)
- [x] refactor to avoid too many `IsUnit` or `nonZeroDivisors`
- [x] split `IsRemainder` and `IsGroebnerBasis`
- [ ] embedding of monomial order (done, not yet submitted), and properties about it and Groebner basis.
- [ ] some theorems can be generalized to avoid requiring `IsUnit`

Dependencies:

- [x] depends on: #26039
- [ ] depends on: #34759

S-polynomial and other lemmas in other files are split from this PR, leaving only changes of `Mathlib/RingTheory/MvPolynomial/Groebner{.lean,/*.lean}` in this PR. The following are PRs split from this PR.

- [x] depends on: #32336
- [x] depends on: #32344
- [x] depends on: #32780
- [x] depends on: #32787
- [x] depends on: #32788
- [x] depends on: #32801
- [x] depends on: #32876
- [x] depends on: #32877
- [ ] depends on: #34873

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Hagb committed:
feat(Mathlib/Data/Finsupp/MonomialOrder): `0 ≤ a` where `a : m.syn`

`simp` would be able to solve it.

Hagb committed:
feat(RingTheory/Ideal/Span.lean): some lemmas about zero and span

`simp` would be able to solve them.

Notice that `span_singleton_zero : span {(0 : R)} = ⊥` itself can be
solved by `simp` without this commit, which uses
`span_singleton_eq_bot : span ({x} : Set α) = ⊥ ↔ x = 0`, while
`simp` cannot solve statements like `f (Ideal.span {(0 : R)}) = f ⊥`,
which would be solvable by `simp` with `span_singleton_zero` marked
with `@[simp]`.

Hagb committed:
Update Mathlib/RingTheory/DedekindDomain/Ideal.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/DedekindDomain/Ideal.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

Hagb committed:
move `dvd_bot` to where it should be

Hagb committed:
feat(Mathlib/RingTheory/MvPolynomial/MonomialOrder): some lemmas about degree

tsuki8 committed:
Merge remote-tracking branch 'origin/mv-degree' into leadingTerm
The theorem need to use the pr submitted by Hagb

tsuki8 committed:
Merge remote-tracking branch 'upstream/MonomialOrder-zero_le' into leadingTerm
need to use zero_le

tsuki8 committed:
feat(RingTheory/MvPolynomial/MonomialOrder): leadingTerm
define the leadingTerm and prove some lemmas related to the def

`leadingTerm`: the leading term of `f` for the monomial ordering `m`

some basic lemmas about leadingTerm including:
1. `leadingTerm_eq_zero_iff`
2. `leadingTerm_image_sdiff_singleton_zero`
3. `leadingTerm_image_insert_zero`
4. `leadingTerm_zero`
5. `leadingTerm_degree_eq`
6. `leadingTerm_degree_eq'`

some lemmas about the degree of `f - m.leadingTerm f`:
1. `degree_sub_leadingTerm`
2. `degree_sub_leadingTerm_lt_degree`
3. `degree_sub_leadingTerm_lt_iff`

tsuki8 committed:
submit `leadingTerm` and related lemmas

tsuki8 committed:
feat(Data/Finsupp/MonomialOrder): add two trival lemma

tsuki8 committed:
change based on the github actions

tsuki8 committed:
Merge remote-tracking branch 'upstream/MonomialOrder-zero_le' into IsRemainder

tsuki8 committed:
Merge remote-tracking branch 'origin/leadingTerm' into IsRemainder

Hagb committed:
add `degree_sum_le`

tsuki8 committed:
Merge remote-tracking branch 'origin/mv-degree' into IsRemainder

Hagb committed:
`lemma` -> `theorem`

Hagb committed:
Merge branch 'master' into Ideal-zeros

Hagb committed:
Update Span.lean

tsuki8 committed:
feat(RingTheory/MvPolynomial/Groebner): define isRemainder and GB

Define isRemainder and Groebner Basis and some relevant lemmas

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
remove unnecessary `classical`

Hagb committed:
Merge commit '4449ce7cca436f2399ffe61ad1546741862144ac' into HEAD

Hagb committed:
improve naming and some proofs

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
fix style

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Riccardo Brasca <riccardo.brasca@gmail.com>

Hagb committed:
Merge remote-tracking branch 'origin/Ideal-zeros' into HEAD

Hagb committed:
Merge remote-tracking branch 'origin/mv-degree' into HEAD

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
Merge remote-tracking branch 'upstream/leadingTerm' into HEAD

Hagb committed:
add more from WuProver/groebner_proj Groebner/Defs.lean

Hagb committed:
remove unnecessary instance argument

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
remove unnecessary imports

Hagb committed:
Apply suggestions from code review

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Apply suggestions from code review

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
add lemmas about ideal

Hagb committed:
Merge remote-tracking branch 'origin/leadingTerm' into HEAD

Hagb committed:
squeeze non-terminal `simp`

Hagb committed:
move `leadingTerm_zero` right after `leadingTerm_zero_iff`

Hagb committed:
update doc and upstreamize lemmas about ideal

Hagb committed:
improve the document of `degree_sub_leadingTerm_lt_iff`

Hagb committed:
Merge branch 'master' into leadingTerm

Hagb committed:
add buchberger theorem as well as some other theorems and lemmas

Hagb committed:
Merge remote-tracking branch 'origin/leadingTerm' into IsRemainder

Hagb committed:
fix some docs and add docs of Groebner basis theorems to header

Hagb committed:
use `insert 0 G` instead of `G ∪ {0}`

Hagb committed:
try to resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2472722695

Hagb committed:
try to resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2472728107

Hagb committed:
add reference

Hagb committed:
use explicit parameter for `MonomialOrder` in def of `sPolynomial`

Hagb committed:
generalize Groebner Basis to allow infinite set

Hagb committed:
update doc of Groebner Basis for the generalization from `Finset` to `Set`

Hagb committed:
add `exists_isGroebnerBasis_finite`

Hagb committed:
generalize `exists_isGroebnerBasis_finite` to `CommRing`

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
generalize `exists_isGroebnerBasis_finite`

Hagb committed:
remove a blank line

Hagb commented on Mathlib/RingTheory/MvPolynomial/Groebner.lean:
```diff
@@ -249,4 +968,636 @@ theorem div_single {b : MvPolynomial σ R}
+  simp_intro .. [or_iff_not_imp_right.mp (hG _ _)]
+
+variable (m) in
+theorem exists_isGroebnerBasis_finite [Finite σ] [IsNoetherianRing R]
```
Hmmm for now I'm also not sure, and I use `IsNoetherianRing (MvPolynomial σ R)` as suggested.

Hagb committed:
remove `degree_sub_leadingTerm{,_lt_degree}` from module doc (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2486569590)

Hagb committed:
remove comment `-- simp?` if the unsqueezed `simp` is short (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2486679733)

chrisflav committed:
simplify `span_leadingTerm_eq_span_monomial` and add `C_mul_leadingCoeff_monomial_degree` (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2486951025)

Hagb committed:
correct naming: `isGroebner` -> `isGroebnerBasis`

Hagb committed:
update doc to describe the way to prove Buchberger Criterion

Hagb committed:
remove duplicated `[Comm{Ring,SemiRing}]` arguments

Hagb committed:
fix indent in mod docstring

Hagb committed:
generate `degree_sub_leadingTerm_lt_degree` with some lemmas added

resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2529881858 and https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2529879468

Hagb committed:
remove `degree_sub_leadingTerm` (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2529882329)

Hagb committed:
Apply suggestion from @erdOne

Co-authored-by: Andrew Yang <36414270+erdOne@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Dagur Asgeirsson <dagurtomas@gmail.com>

Hagb committed:
Merge branch 'master' into leadingTerm

Hagb committed:
Merge remote-tracking branch 'origin/leadingTerm' into HEAD

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jcommelin commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -611,6 +613,16 @@ theorem degree_prod [IsCancelMulZero R] {ι : Type*} {P : ι → MvPolynomial σ
   rw [leadingCoeff_ne_zero_iff]
   exact H i hi
 
+lemma degree_mul' [IsCancelMulZero R] [NoZeroDivisors R]
```
The changes in this file could be split into a separate PR, for ease of reviewing.

Hagb committed:
feat(Algebra/GroupWithZero/NonZeroDivisors): add some lemmas about product

Hagb committed:
use `⁰` macro instead of raw `nonZeroDivisors`

Hagb committed:
feat(RingTheory/MvPolynomial/MonomialOrder): generate some `IsRegular` hypotheses to `∈ nonZeroDivisors`

Hagb committed:
`degree_smul` -> `degree_smul_of_isRegular`

Hagb committed:
feat(RingTheory/MvPolynomial/MonomialOrder): add several lemmas about `leadingTerm` and inequality on degree

Hagb committed:
feat(RingTheory/MvPolynomial/MonomialOrder): add some variants of `sPolynomial_mul_monomial`

Hagb committed:
feat(RingTheory/MvPolynomial/Ideal): add lemmas that `sPolynomial` preserves ideal membership

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
Merge remote-tracking branch 'origin/MonomialOrder-nonZeroDivisors' into HEAD

Hagb committed:
Merge remote-tracking branch 'origin/leadingTerm-lemmas' into HEAD

Hagb committed:
Merge remote-tracking branch 'origin/sPolynomial_mul-variants' into HEAD

Hagb committed:
clean up the proof of Buchberger Criterion

Hagb committed:
generalize from regular to non-zero-divisor

Hagb committed:
Merge remote-tracking branch 'origin/sPolynomial-ideal' into HEAD

Hagb committed:
split into several files

Hagb committed:
feat(Algebra/MvPolynomial/Basic): add `MvPolynomial.support_C`

Hagb committed:
feat(Data/Set/Operations): add two lemmas about equality of `rangeFactorization`

Hagb committed:
Merge remote-tracking branch 'origin/rangeFactorization' into HEAD

Hagb committed:
Merge remote-tracking branch 'origin/support_C' into HEAD

Hagb committed:
clean up and golf proofs, and squeeze `simp`

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
fix some errors in the informal proof of Buchberger's Criterion

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
feat(Algebra/Order/Monoid/Unbundled/WithTop): `AddEquiv` between `WithBot`

Hagb committed:
feat(Logic/Basic): `a ∧ (a → b) ↔ a ∧ b` and `(a → b) ∧ a ↔ b ∧ a`

Hagb committed:
feat(RingTheory/MvPolynomial/MonomialOrder): misc lemmas

Hagb committed:
feat({RingTheory/MvPolynomial,Data/Finsupp}/MonomialOrder): add degree with ⊥ as degree of 0

Hagb committed:
update Groebner basis

- refactor something with `WithBot (σ →₀ ℕ)` variant of `MonomialOrder.degree`
- refactor to use namespaces `MonomialOrder.{IsGroebnerBasis,IsRemainder}` instead of `*_of_{isGroebnerBasis,isRemainder}` in theorem names.
- refactor to avoid `nonZeroDivisors` as hypothesis

Hagb committed:
Merge remote-tracking branch 'origin/mv-misc' into HEAD

Hagb committed:
Merge remote-tracking branch 'origin/prop-lemma' into HEAD

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
fix error

Hagb committed:
typo: `WithBot.withBotCongr` -> `Equiv.withBotCongr`

Hagb committed:
Merge remote-tracking branch 'origin/withBotCongr' into HEAD

Hagb committed:
remove a lemma about `rename` to avoid depending on #34757

Hagb committed:
Merge remote-tracking branch 'origin/withBotDegree' into HEAD

Hagb committed:
characterize `IsRemainder` with `IsRemainder` on finset

Hagb committed:
style: def -> lemma

Hagb committed:
fix simp

Hagb committed:
Merge remote-tracking branch 'origin/withBotDegree' into HEAD

Hagb committed:
feat(Data/Finsupp/Basic): add `comapDomain_embDomain` and generalize `comapDomain_surjective`

The proof of `comapDomain_embDomain` is modified from proof of
`comapDomain_mapDomain`.

Hagb committed:
Merge remote-tracking branch 'origin/comapDomain_surjective' into HEAD

Hagb committed:
golf `isRemainder_insert_zero_iff_isRemainder`
